swarm/pss/pss.go

```diff
@@ -241,91 +302,244 @@ func (self *Pss) checkFwdCache(addr []byte, digest pssDigest) bool {
 	return false
 }
 
-func (self *Pss) getHandlers(topic Topic) map[*Handler]bool {
+func (self *Pss) getHandlers(topic whisper.TopicType) map[*Handler]bool {
 	self.lock.Lock()
 	defer self.lock.Unlock()
 	return self.handlers[topic]
 }
 
+// protocol handler for incoming pss msg
+// check if address partially matches = CAN be for us
 func (self *Pss) handlePssMsg(msg interface{}) error {
-	pssmsg := msg.(*PssMsg)
+	pssmsg, ok := msg.(*PssMsg)
+	if ok {
+		if !self.isSelfPossibleRecipient(pssmsg) {
+			log.Trace("pss was for someone else :'( ... forwarding")
+			return self.Forward(pssmsg)
+		}
+		log.Trace("pss for us, yay! ... let's process!")
 
-	if !self.isSelfRecipient(pssmsg) {
-		log.Trace("pss was for someone else :'( ... forwarding")
-		return self.Forward(pssmsg)
+		return self.Process(pssmsg)
 	}
-	log.Trace("pss for us, yay! ... let's process!")
-	return self.Process(pssmsg)
+
+	return fmt.Errorf("invalid message")
 }
 
-// Entry point to processing a message for which the current node is the intended recipient.
+// Entry point to processing a message for which the current node can be the intended recipient.
 func (self *Pss) Process(pssmsg *PssMsg) error {
+	var recvmsg *whisper.ReceivedMessage
+	var from pot.Address
 	env := pssmsg.Payload
-	payload := env.Payload
+
 	handlers := self.getHandlers(env.Topic)
 	if len(handlers) == 0 {
 		return fmt.Errorf("No registered handler for topic '%x'", env.Topic)
 	}
-	nid, _ := discover.HexID("0x00")
-	p := p2p.NewPeer(nid, fmt.Sprintf("%x", env.From), []p2p.Cap{})
-	for f := range handlers {
-		err := (*f)(payload, p, env.From)
+
+	// try all our symkeys in order
+	// we could should these (by last used first?) to possibly match quicker
+	if len(env.AESNonce) > 0 { // see whisper.envelope.go.OpenSymmetric
+		//for _, symkeyid := range self.recvsymkeys {
+		for symkeyid, _ := range self.reverseKeyPool {
+			log.Trace("attempting symmetric decrypt", "symkey", symkeyid)
+			symkey, err := self.w.GetSymKey(symkeyid)
+			if err == nil {
+				recvmsg, err = env.OpenSymmetric(symkey)
+				if err == nil {
+					if !recvmsg.Validate() {
```

Comment:
We should avoid the indentation by reversing the conditions and using `continue`:

```go
symkey, err := self.w.GetSymKey(symkeyid)
if err != nil {
	// log the error
	continue
}
recvmsg, err := env.OpenSymmetric(symkey)
if err != nil {
	// log the error
	continue
}
...
```